### PR TITLE
INT-3661: Fix the eager BF access from BPPs (P I)

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3661
## Status Quo

Previously there were a lot of noise from the `PostProcessorRegistrationDelegate$BeanPostProcessorChecker` for early access for beans.
That may produce some side-effects when some of `BeanFactoryPostProcessor`s won't adjust those beans.

The issue is based on two facts:
1. Loading beans from `BPP`, e.g. `IntegrationEvaluationContextAwareBeanPostProcessor` (or `ChannelSecurityInterceptorBeanPostProcessor` - https://jira.spring.io/browse/INT-3663)
2. Loading beans from `setBeanFactory()/setApplicationContext()` container methods
## The fix
- Move all stuff from `setBeanFactory()` with access to the `BeanFactory` (e.g. `this.messageBuilderFactory = IntegrationUtils.getMessageBuilderFactory(this.beanFactory);`)
  to some other lazy-load methods like `getMessageBuilderFactory()`
- Fix parser tests to `assertNotSame` for `messageBuilderFactory` since there is no activity for target components to lazy-load the stuff
- Polish some test according the new lazy-load logic
- Rework `IntegrationEvaluationContextAwareBeanPostProcessor` to the `SmartInitializingSingleton` and make it `Ordered`
- Populate `beanFactory` for the internal instance of `connectionFactory` in the `TcpSyslogReceivingChannelAdapter`
- Populate `beanFactory` for the internal `UnicastReceivingChannelAdapter` in the `UdpSyslogReceivingChannelAdapter`
- Add `log.info` that `UdpSyslogReceivingChannelAdapter` overrides `outputChannel` for the provided `UnicastReceivingChannelAdapter`
- Change the internal `MessageChannel` in the `UdpSyslogReceivingChannelAdapter` to the `FixedSubscriberChannel` for better performance

### CODE REVIEW

**Comment by artembilan** [comment]
Also fixes https://jira.spring.io/browse/INT-3662

File: spring-integration-core/src/main/java/org/springframework/integration/util/AbstractExpressionEvaluator.java [hunk 1]
```diff
@@ -54,7 +54,9 @@
 
 	private volatile BeanFactory beanFactory;
 
-	private volatile MessageBuilderFactory messageBuilderFactory;
+	private volatile MessageBuilderFactory messageBuilderFactory = new DefaultMessageBuilderFactory();
+
+	private volatile boolean messageBuilderFactorySet;
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/util/AbstractExpressionEvaluator.java]: This class has a bug in the `afterPropertiesSet()` will be fixed tomorrow

**Comment by artembilan** [comment]
Pushed

**Comment by artembilan** [comment]
Pushed fix for `MongoDbMessageStoreClaimCheckIntegrationTests`. I haven't had a local MongoDB before, hence all tests have been bypassed by Rule reason :smile:

File: spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java [hunk 2]
```diff
@@ -57,7 +58,7 @@ public void interceptor() {
 				TestUtils.getPropertyValue(channel, "dispatcher"), "maxSubscribers", Integer.class).intValue());
 		channel = context.getBean("pubSub", MessageChannel.class);
 		Object mbf = context.getBean(IntegrationUtils.INTEGRATION_MESSAGE_BUILDER_FACTORY_BEAN_NAME);
-		assertSame(mbf, TestUtils.getPropertyValue(channel, "dispatcher.messageBuilderFactory"));
+		assertNotSame(mbf, TestUtils.getPropertyValue(channel, "dispatcher.messageBuilderFactory"));
```
Comment (garyrussell) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java]: This makes no sense to me; the goal was to use the global MBF for __all__ message builds.

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractAggregatingMessageGroupProcessor.java [hunk 3]
```diff
@@ -64,11 +79,12 @@ public final Object processMessageGroup(MessageGroup group) {
 		Map<String, Object> headers = this.aggregateHeaders(group);
 		Object payload = this.aggregatePayloads(group, headers);
 		AbstractIntegrationMessageBuilder<?> builder;
+		MessageBuilderFactory messageBuilderFactory = getMessageBuilderFactory();
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractAggregatingMessageGroupProcessor.java]: We don't need to store in a local variable here; it's only used once on each branch of the `if`.

File: spring-integration-core/src/main/java/org/springframework/integration/aop/MessagePublishingInterceptor.java [hunk 4]
```diff
@@ -163,9 +173,10 @@ private void publishMessage(Method method, StandardEvaluationContext context) th
 		Expression expression = this.parser.parseExpression(payloadExpressionString);
 		Object result = expression.getValue(context);
 		if (result != null) {
+			MessageBuilderFactory messageBuilderFactory = getMessageBuilderFactory();
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aop/MessagePublishingInterceptor.java]: Same here

File: spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationObjectSupport.java [hunk 5]
```diff
@@ -143,9 +142,15 @@ public void setChannelResolver(DestinationResolver<MessageChannel> channelResolv
 
 	@Override
 	public final void afterPropertiesSet() {
+		this.integrationProperties = IntegrationContextUtils.getIntegrationProperties(this.beanFactory);
 		try {
 			if (this.messageBuilderFactory == null) {
-				this.messageBuilderFactory = IntegrationUtils.getMessageBuilderFactory(this.beanFactory);
+				if (this.beanFactory != null) {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationObjectSupport.java]: Shouldn't we just move this to `getMessageBuilderFactory()` to make it really lazy.

File: spring-integration-core/src/main/java/org/springframework/integration/expression/IntegrationEvaluationContextAwareBeanPostProcessor.java [hunk 6]
```diff
@@ -50,4 +58,18 @@ public Object postProcessBeforeInitialization(Object bean, String beanName) thro
 	public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
 		return bean;
 	}
+
+	@Override
+	public void afterSingletonsInstantiated() {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/expression/IntegrationEvaluationContextAwareBeanPostProcessor.java]: Nice!

File: spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java (hunk 2 above)
Comment (artembilan) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java]: Right. I understand your original goal. But I wanted to pay your attention here that we now is in the lazy-load, hence these tests really don't make sense because there is no anymore any interaction.
Are you OK if I remove these `assertNotSame` with the next commit/polishing to this PR ?

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractAggregatingMessageGroupProcessor.java (hunk 3 above)
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractAggregatingMessageGroupProcessor.java]: OK. No problem! IDEA side-effect :smile:

File: spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationObjectSupport.java (hunk 5 above)
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationObjectSupport.java]: Will be done

File: spring-integration-core/src/main/java/org/springframework/integration/expression/IntegrationEvaluationContextAwareBeanPostProcessor.java (hunk 6 above)
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/expression/IntegrationEvaluationContextAwareBeanPostProcessor.java]: Yes, but pay attention to the  `IntegrationEvaluationContextAware` JavaDocs, please.

File: spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java (hunk 2 above)
Comment (garyrussell) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java]: Oh, I see; you're not saying that at runtime we're using a different builder just that, since these are parser only tests, we won' have it yet.

I agree; they should just be removed.

However, we might want to consider adding verification to runtime tests instead :smile: 

(That can be a separate PR if you prefer)

File: spring-integration-core/src/main/java/org/springframework/integration/expression/IntegrationEvaluationContextAwareBeanPostProcessor.java (hunk 6 above)
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/expression/IntegrationEvaluationContextAwareBeanPostProcessor.java]: I'm ok with that;

File: spring-integration-core/src/main/java/org/springframework/integration/util/AbstractExpressionEvaluator.java [hunk 7]
```diff
@@ -82,16 +81,13 @@ public void setConversionService(ConversionService conversionService) {
...
 		return this.messageBuilderFactory;
 	}
 
 	@Override
 	public void afterPropertiesSet() throws Exception {
 		getEvaluationContext();
-		if (this.messageBuilderFactory == null) {
+		if (this.beanFactory != null) {
 			this.messageBuilderFactory = IntegrationUtils.getMessageBuilderFactory(this.beanFactory);
 		}
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/util/AbstractExpressionEvaluator.java]: Should we move this to `evaluateExpression` ?

File: spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/store/JdbcChannelMessageStore.java [hunk 8]
```diff
@@ -392,6 +394,9 @@ public void afterPropertiesSet() throws Exception {
 			logger.warn("The jdbcTemplate's fetchsize is not 1. This may cause FIFO issues with Oracle databases.");
 		}
 
+		if (this.beanFactory != null) {
+			this.messageBuilderFactory = IntegrationUtils.getMessageBuilderFactory(this.beanFactory);
```
Comment (garyrussell) [inline spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/store/JdbcChannelMessageStore.java]: Lazy?

Comment (garyrussell) [inline spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/store/JdbcChannelMessageStore.java]: Actually, I am not sure these need to be lazy at all; we're fetching 1 bean which has no other deps.

Shouldn't we just remove the usages in setBeanFactory() and move to `afterPropertiesSet()` instead of lazy loading?

**Comment by garyrussell** [comment]
Wow - a lot of repetitive, thankless work (but I thank you !!).

**Comment by garyrussell** [comment]
Nothing more from me.

File: spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java (hunk 2 above)
Comment (artembilan) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests.java]: Yes, I also intend to remove them. 
Re. separate JIRA: no need, really. See `PublishSubscribeChannelParserTests` changes. I hope that should prove that the fix works as promised.

File: spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/store/JdbcChannelMessageStore.java (hunk 8 above)
Comment (artembilan) [inline spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/store/JdbcChannelMessageStore.java]: To be honest: I followed **exactly** that rule. If there is an `InitializingBean`, I did that in the `afterPropertiesSet()`, otherwise lazy `getMessageBuilderFactory()`.
Have I missed anything ?

**Comment by artembilan** [comment]
Pushed

**Comment by garyrussell** [comment]
Merged as 2bde14b742d55b69ffc64bcef3aefce114cab557
